tests/integration/relation_tests/legacy_relations/helpers.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+import logging
+import time
+from typing import Optional
+
+import yaml
+from pytest_operator.plugin import OpsTest
+
+from tests.integration.relation_tests.legacy_relations.api import GraylogApi
+
+logger = logging.getLogger(__name__)
+DEFAULT_REST_API_TIMEOUT = 120
+GRAYLOG_PORT = 9000
+GRAYLOG_APP_NAME = "graylog"
+
+
+async def get_application_relation_data(
+    ops_test: OpsTest, application_name: str, key: str, related_unit_name: str
+) -> Optional[str]:
+    """Get relation data for an application.
+
+    Args:
+        ops_test: The ops test framework instance
+        application_name: The name of the application
+        key: key of data to be retrieved
+
+    Returns:
+        the that that was requested or None
+            if no data in the relation
+
+    Raises:
+        ValueError if it's not possible to get application unit data.
+    """
+    related_app = related_unit_name.split("/")[0]
+    unit_name = f"{application_name}/0"
+    raw_data = (await ops_test.juju("show-unit", unit_name))[1]
+    if not raw_data:
+        raise ValueError(f"no unit info could be grabbed for {unit_name}")
+
+    data = yaml.safe_load(raw_data)
+
+    for relation in data[unit_name]["relation-info"]:
+        if relation["endpoint"] == related_app:
+            return relation["related-units"][related_unit_name]["data"][key]
+
+
+async def get_graylog_client(ops_test: OpsTest):
+    unit = ops_test.model.applications[GRAYLOG_APP_NAME].units[0]
+    base_url = "http://{}:{}/api/".format(unit.public_address, GRAYLOG_PORT)
+    return GraylogApi(
+        base_url=base_url,
+        username="admin",
+        password=await get_password(ops_test, GRAYLOG_APP_NAME),
+        token_name="graylog-charm",
+    )
+
+
+async def _verify_rest_api_is_alive(ops_test: OpsTest, timeout=DEFAULT_REST_API_TIMEOUT):
+    g = await get_graylog_client(ops_test)
+    url = ""  # Will query using the base URL of the client, i.e. /api/
+    resp = g.request(url)
+    start_ts = time.time()
+    while resp is None:
+        time.sleep(5)
+        resp = g.request(url)
+        if time.time() - start_ts > timeout:
+            raise ApiTimeoutError()
```

Comment:
Maybe you can use `tenacity` here? I'm not sure if it can replace this logic completely.

Reply:
good point! This was some code I copied from the original graylog, must have been written before tenacity 😅